### PR TITLE
Add profitability check to array-property-opt

### PR DESCRIPTION
**Pull request by meg-gupta**
This change adds additional heuristics to array-property-opt to avoid code size increase in cases where the optimization may not be beneficial. With this change, we do not specialize a loop if it has any opaque function calls or the instruction count computed recursively exceeds a predefined threshold.

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci Please Test Source Compatibility

**Comment by meg-gupta** [comment]
@swift-ci Please benchmark

**Comment by gottesmm** [comment]
@meg-gupta you also want to do the perf suite. Can give you code-size #s

**Comment by gottesmm** [comment]
@swift-ci test compiler performance

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp [hunk 1]
```diff
@@ -107,26 +150,55 @@ class ArrayPropertiesAnalysis {
...
+
+    for (auto *BB : Loop->getBlocks()) {
+      for (auto &Inst : *BB) {
+        if (InstCount++ > MaxInstThreshold) {
+          return false;
+        }
+
+        if (auto Apply = FullApplySite::isa(&Inst)) {
+          const auto Callee = Apply.getReferencedFunctionOrNull();
+          auto CalleeInstCount = checkProfitabilityRecursively(Callee);
```
Comment (eeckstein) [inline lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: I'm a bit concerned about the complexity. You're scanning a potentially large part of the whole module for _every_ function on which this pass runs. This is quadratic.
But it might be no problem if you reduce MaxInstThreshold.
Alternatively, you could make a simple Analysis, which contains the InstCountCache. This would cache the instruction count for functions across multiple calls of this pass.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp [hunk 2]
```diff
@@ -107,26 +150,55 @@ class ArrayPropertiesAnalysis {
     // beneficial. This heuristic also simplifies which regions we want to
     // specialize on. We will specialize the outermost loopnest that has
     // 'array.props' instructions in its preheader.
+
```
Comment (eeckstein) [inline lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: Did you do this reformatting on purpose? If yes, you might want to put it into a separate commit, e.g. with a message "Improve formatting [...] NFC". Just to make it clear.

**Review by eeckstein — CHANGES_REQUESTED** [review]

**Comment by eeckstein** [comment]
Forgot to mention: can you add tests? E.g. including the case where a recursive function is called within the loop.
For better testing you could make the MaxInstThreshold configurable with a command line option and set it to a very low limit just for the test.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 2 above)
Comment (meg-gupta) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: This wasn't on purpose. I'll get rid of it. Its left-over from undoing the clang-format.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp [hunk 3]
```diff
@@ -82,15 +82,58 @@ class ArrayPropertiesAnalysis {
...
+    uint32_t InstCount = 0;
+
+    for (auto &BB : *Callee) {
+      for (auto &I : BB) {
+        if (InstCount++ > MaxInstThreshold) {
+          InstCountCache[Callee] = MaxInstThreshold;
+          return MaxInstThreshold;
+        }
+        if (auto Apply = FullApplySite::isa(&I)) {
+          const auto CalleeInstCount = checkProfitabilityRecursively(
```
[the remark this answers is not in the document]
Comment (meg-gupta) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: I have handled the recursive case. I add an entry with count 0 just above.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp [hunk 4]
```diff
@@ -82,15 +82,58 @@ class ArrayPropertiesAnalysis {
...
+    if (!Callee) {
+      return MaxInstThreshold;
+    }
+
+    auto CacheEntry = InstCountCache.find(Callee);
+    if (CacheEntry != InstCountCache.end()) {
+      return CacheEntry->second;
+    }
+
+    InstCountCache.insert(std::make_pair(Callee, 0));
```
Comment (meg-gupta) [inline lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: @eeckstein adding the entry to InstCountCache to avoid cycles here

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp [hunk 5]
```diff
@@ -82,15 +82,58 @@ class ArrayPropertiesAnalysis {
...
 
+  llvm::DenseMap<SILFunction *, uint32_t> InstCountCache;
   llvm::SmallSet<SILValue, 16> HoistableArray;
 
   SmallPtrSet<SILBasicBlock *, 16> ReachingBlocks;
   SmallPtrSet<SILBasicBlock *, 16> CachedExitingBlocks;
+
+  // This controls the max permissible instruction count of the loop for
+  // specialization
+  const uint32_t MaxInstThreshold = 5000;
```
[the remark this answers is not in the document]
Comment (meg-gupta) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: I probably should have used a better name. 5000 also includes the instruction count of the callee functions summed up with the loop's instruction count.  I analyzed one benchmark, and it looks like mostly small loops qualify for this optimization after the legality checks. But they do call a bunch of library functions that can have instruction counts unto 1000. 

And am yet to run some more experiments with the 5000 threshold.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 1 above)
Comment (meg-gupta) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: Ok. I was worried about the cache not being fully utilized as well. I'll see how this can be improved. Thanks for the suggestion.

**Comment by meg-gupta** [comment]
Leaving this PR on hold for a bit. I am investigating ways we improve analyses for LICM, which can avoid the need for array-property-opt in some cases.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 2 above)
Comment (atrick) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: LLVM style is to omit curly brackets for single line conditional clauses. I have a `git-format` script that does this:
`git clang-format --binary ${CLANG_FORMAT} --style "{BasedOnStyle: LLVM, BreakBeforeBinaryOperators: NonAssignment}"`

The break-before-binary operators is something I do to avoid bugs in compound multiline expressions where, if the operator is hidden at the end of the line in some arbitrary column amidst the noise, my brain interpolates the operator I expect so otherwise I never spot those bugs.

Comment (gottesmm) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: @atrick if we want to use that style, we should add it to Swift's default clang-format.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 1 above)
Comment (atrick) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: @meg-gupta @eeckstein 
Here are some suggestions:

We could have a general bottom-up inter-procedural analysis for "estimated function computation complexity". It would still need a threshold though, or the analysis will be much more expensive than what you're already doing. The cost to compute it is `num_functions * threshold`, where the threshold is reached instantly at any external, virtual, or witness call (it should not follow all known callees). That should be very cheap in practice. But, we should also have a multiplier for the level of the instruction in the loop nest, which means computing loops.

I think this is overkill for just this one purpose, but it seems like a generally useful thing to have. We should find out if others agree before spending time on it.

Just to kill this argument before someone brings it up... I do *not* think this should be combined with the algorithm for inlining cost analysis. None of the inliner's complexity is relevant to this metric. It's designed around processing blocks and instructions in a particular order to propagate constants. And it's determining code size cost, not computation complexity. Reusing the logic for unrelated metrics would be confusing. However, `instructionInlineCost` is currently simple enough that it could be reused for both code size and computation cost of a single instruction at a time. It essentially allows you to ignore most SIL instructions as having no impact on generated code.

Comment (eeckstein) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: Yeah, a general bottom-up analysis sounds good for this purpose.
And sorry, I have to bring this up: the ShortestPathAnalysis of the inliner serves a very similar purpose. So it would be cool if we could extract the ShortestPathAnalysis from the inliner and make it a general tool which can also be used here.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 2 above)
Comment (atrick) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: @gottesmm The last discussion about changing the default ended with no one opposing this variation in style, most people agreeing that this was within the current guidelines, but almost everyone opposing having a discussion about changing the default. For me, it's a very real issue of the quality of the code I work on. It's got nothing to do with making things pretty.

Comment (gottesmm) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: That isn't what I meant. I meant put up a PR. I am fine with the change = ).

**Comment by airspeedswift** [comment]
@swift-ci test compiler performance

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 2 above)
Comment (meg-gupta) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: clang-format is reordering more than what I changed. Are such changes generally ok ?

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (meg-gupta) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp]: The AnalysisThreshold is 5000 and LoopInstThreshold is 500 in my latest commit.

**Comment by meg-gupta** [comment]
@swift-ci Test Source Compatibility
@swift-ci please benchmark
@swift-ci test compiler performance

**Comment by theblixguy** [comment]
Unfortunately passing multiple CI commands in a single comment doesn’t work, so you need to pass one at a time.

**Comment by meg-gupta** [comment]
@swift-ci Test Source Compatibility

**Comment by meg-gupta** [comment]
@swift-ci please benchmark

**Comment by meg-gupta** [comment]
@swift-ci test compiler performance

**Comment by meg-gupta** [comment]
Regression | OLD | NEW | DELTA | RATIO
-- | -- | -- | -- | --
MapReduceNSDecimalNumber | 114 | 291 | +155.3% | 0.39x
MapReduceNSDecimalNumberShort | 279 | 458 | +64.2% | 0.61x
Set.isSubset.Seq.Box0 | 1152 | 1481 | +28.6% | 0.78x
Set.isSubset.Seq.Box25 | 1306 | 1633 | +25.0% | 0.80x
Set.intersection.Seq.Box0 | 365 | 432 | +18.4% | 0.84x
Set.intersection.Seq.Box25

MapReduce test regressions are because the control flow pattern changes limit some jump threading which in turn limit arc optimizations. The regression is due to additional obj.retain/release calls in the loop.

SetTest regressions are because it sees a library function Set.Variant.insert which has a class_method that cannot be devirtualized, even otherwise the AnalysisThreshold will exceed scanning the library function call chain.

**Review by atrick — APPROVED** [review]
I have no concerns with checking this in if it meets our immediate needs. The implementation looks good. It's clear how it could be extended and generalized in the future if that becomes worthwhile.

If you're not immediately fixing those regressions, then please file a bug to follow up!

**Comment by meg-gupta** [comment]
@swift-ci smoke test please

**Comment by meg-gupta** [comment]
@swift-ci Please smoke test

**Review by eeckstein — APPROVED** [review]
lgtm
